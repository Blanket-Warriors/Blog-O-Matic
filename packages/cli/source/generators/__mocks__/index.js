export const postGenerator = jest.fn()
export const blogGenerator = jest.fn()
