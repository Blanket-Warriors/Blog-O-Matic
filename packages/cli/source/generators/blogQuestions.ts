export const basicQs = [
  {
    message: "What is your name?",
    name: "author",
    type: "input",
  },
  {
    default: "my-blog",
    message: "What is the title of this blog?",
    name: "title",
    type: "input",
  },
  {
    default: "YYYY/MM/DD HH:MM",
    message: "How do you like to write your dates?",
    name: "dateFormat",
    type: "input",
  },
  {
    choices: [ "scp", "fs" ],
    message: "How would you like to publish your blog?",
    name: "publisher",
    type: "list",
  },
]

export const scpQs = [
  {
    default: "example.com",
    message: "Host: ",
    name: "host",
    type: "input",
  },
  {
    default: "",
    message: "Port: ",
    name: "port",
    type: "input",
  },
  {
    default: "",
    message: "User: ",
    name: "user",
    type: "input",
  },
  {
    default: "~/documents/my-blog",
    message: "Path on server:",
    name: "path",
    type: "input",
  },
  {
    default: true,
    message: "Will you be using an ssh key for access?",
    name: "ssh",
    type: "confirm",
  },
]

export const sshQs = [
  {
    default: "~/.ssh/id_rsa",
    message: "Where is your private key located?",
    name: "ssh",
    type: "input",
  },
]

