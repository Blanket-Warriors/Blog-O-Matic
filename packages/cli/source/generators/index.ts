export * from "./blogGenerator"
export * from "./postGenerator"
