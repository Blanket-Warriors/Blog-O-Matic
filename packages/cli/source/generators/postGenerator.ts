import { get, promisify } from "@civility/utilities"
import * as fs from "fs"
import { prompt } from "inquirer"
import { safeLoad } from "js-yaml"
import { join } from "path"
import { IPost } from "../definitions"
import { postTemplate } from "../templates"
const writeFile = promisify(fs.writeFile)


export async function postGenerator() {
  console.log("\nðŸ˜³ðŸ˜³ðŸ¤–ðŸ˜³ LET'S MAKE A BLOG POST! ðŸ˜³ðŸ˜³ðŸ¤–ðŸ˜³\n")

  const config = safeLoad(fs.readFileSync("./blog.config.yml", "utf8")) || {}

  const questions = [
    {
      default: new Date().toLocaleString(undefined, {
        day: "numeric",
        month: "long",
        year: "numeric",
      }),
      message: "title:",
      name: "title",
      type: "input",
    },
    {
      default: get(config, [ "blog", "author" ]) || "",
      message: "author:",
      name: "author",
      type: "input",
    },
    {
      default: get(config, [ "blog", "location" ]) || "",
      message: "location:",
      name: "location",
      type: "input",
    },
  ]

  const { author = "", location = "", title = "" } = await prompt(questions) as IPost
  const filename = title.replace(",", "").replace(/[^a-zA-Z0-9_.@()-]/g, "-").toLowerCase()
  await writeFile(join("posts", `${filename}.md`), postTemplate({
    author, location, title,
    CREATED: new Date().getTime(),
    UPDATED: new Date().getTime(),
  }))

  console.log(`
    Congratulations! ðŸŽ‰ ðŸŽ‰ ðŸŽ‰
    You generated a blog post!
  `)
}
