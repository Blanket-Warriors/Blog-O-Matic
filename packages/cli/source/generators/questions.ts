export const blogAuthor = {
  message: "What is your name?",
  name: "author",
  type: "input",
}

export const blogTitle = {
  default: "my-blog",
  message: "What is the title of this blog?",
  name: "title",
  type: "input",
}

export const blogOut = {
  default: "./build",
  message: "Where would you like build files to be?",
  name: "out",
  type: "input",
}

export const blogPublisher = {
  choices: [ "fs", "s3", "scp" ],
  message: "How would you like to publish your blog?",
  name: "publisher",
  type: "list",
}

export const host = {
  default: "127.0.0.1",
  message: "What is the remote server's hostname?",
  name: "host",
  type: "input",
}

export const path = {
  default: "~/documents/my-blog",
  message: "Where is the location on the remote server?",
  name: "path",
  type: "input",
}

export const port = {
  message: "What is the remote server's port? (optional)",
  name: "port",
  type: "input",
}

export const postTitle = {
  default: new Date().toLocaleString(undefined, {
    day: "numeric",
    month: "long",
    year: "numeric",
  }),
  message: "title:",
  name: "title",
  type: "input",
}

export const postAuthor = {
  message: "Who wrote this post?",
  name: "author",
  type: "input",
}

export const s3Bucket = {
  message: "What is the name of your bucket?",
  name: "bucket",
  type: "input",
}

export const s3Creds = {
  default: "~/documents/rootkey.csv",
  message: "Where is your credentials file?",
  name: "creds",
  type: "input",
}

export const user = {
  message: "What is the remote server's user? (optional)",
  name: "path",
  type: "input",
}
