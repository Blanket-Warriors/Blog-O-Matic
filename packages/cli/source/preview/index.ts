import { startServer } from "@blog-o-matic/editor"

export const preview = (tempDir: string) => {
  startServer(tempDir)
}
