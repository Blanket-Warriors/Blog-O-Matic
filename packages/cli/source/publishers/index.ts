export * from "./fsPublisher/fsPublisher"
export * from "./s3Publisher/s3Publisher"
