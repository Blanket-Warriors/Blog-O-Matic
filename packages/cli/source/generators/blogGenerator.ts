import { promisify } from "@civility/utilities"
import * as fs from "fs"
import { createPromptModule } from "inquirer"
import { join } from "path"
import { configTemplate, ignoreTemplate, privateTemplate } from "../templates"
import { basicQs, scpQs, sshQs } from "./blogQuestions"
const prompt = createPromptModule()
const homedir = require("os").homedir()
const mkdir = promisify(fs.mkdir)
const writeFile = promisify(fs.writeFile)


export interface IConfig {
  AUTHOR: string
  DATE_FORMAT: string
  TITLE: string
  SERVICE: string
}

export interface IPrivate {
  FILESYSTEM_LOCATION?: string
  PORT?: string
  SERVER_LOCATION?: string
  SSH_LOCATION?: string
  USE_SSH?: boolean
}


export async function blogGenerator() {
  console.log("\nðŸ˜³ðŸ˜³ðŸ¤–ðŸ˜³ WELCOME TO BLOG-O-MATIC! ðŸ˜³ðŸ˜³ðŸ¤–ðŸ˜³\n")

  const config = await prompt(basicQs) as IConfig
  const privateInfo: IPrivate | null = config.SERVICE === "scp"
    ? await prompt(scpQs) as IPrivate
    : null

  if (privateInfo && privateInfo.USE_SSH) {
    const LOCATION_PROMPTS = await prompt(sshQs) as { SSH_LOCATION: string }
    privateInfo.SSH_LOCATION = LOCATION_PROMPTS.SSH_LOCATION
  }

  try {
    await Promise.all([
      mkdir(config.TITLE),
      mkdir(join(homedir, ".blog-o-matic"), { recursive: true }),
    ])
  } catch (error) {
    if (error.code === "EEXIST") return console.warn("FAILED: A blog already exists here!")
  }

  await Promise.all([
    writeFile(join(config.TITLE, "blog.config.yml"), configTemplate(config)),
    writeFile(join(config.TITLE, ".blogignore"), ignoreTemplate()),
    mkdir(join(config.TITLE, "resources")),
    mkdir(join(config.TITLE, "posts")),
  ])

  if (privateInfo) {
    writeFile(join(homedir, ".blog-o-matic", `${config.TITLE}.yml`), privateTemplate(privateInfo))
  }

  console.log(`
    Congratulations! ðŸŽ‰ ðŸŽ‰ ðŸŽ‰
    Blog-o-Matic successfully created your edit folder!
    You can write whatever you want in here!
    (Or if you're lazy like us, you can run \`blog post\`)
  `)
}
