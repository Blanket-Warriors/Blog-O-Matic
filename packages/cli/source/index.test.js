jest.mock("fs")
jest.mock("js-yaml", () => ({ load: jest.fn(() => ({ publisher: "fs" })) }))
jest.mock("path")
jest.mock("./generators")
jest.mock("./preview")
jest.mock("./publishers")


import * as cli from "./index"
import { blogGenerator, postGenerator } from "./generators"
import { preview } from "./preview"
import { fsPublisher, scpPublisher } from "./publishers"


beforeAll(() => { console.log = jest.fn() })
beforeEach(() => jest.clearAllMocks())


test("Should export public API", () => {
  expect(typeof cli.blogGenerator).toBe("function")
  expect(typeof cli.postGenerator).toBe("function")
  expect(typeof cli.preview).toBe("function")
  expect(typeof cli.fsPublisher).toBe("function")
  expect(typeof cli.scpPublisher).toBe("function")
  expect(typeof cli.start).toBe("function")
})


test("Should log help text on no argument", () => {
  cli.start()
  expect(blogGenerator).toBeCalledTimes(0)
  expect(postGenerator).toBeCalledTimes(0)
  expect(preview).toBeCalledTimes(0)
  expect(fsPublisher).toBeCalledTimes(0)
  expect(scpPublisher).toBeCalledTimes(0)

  expect(console.log).toBeCalledTimes(1)
  expect(console.log.mock.calls[0][0]).toMatchSnapshot()
})

test("Should only run one command at a time", () => {
  cli.start("init")
  cli.start("post")
  cli.start("publish")
  expect(preview).toBeCalledTimes(0)
})

test("Should run blog generator on `blog init`", () => {
  cli.start("init")
  expect(blogGenerator).toBeCalledTimes(1)
})

test("Should run post generator on `blog post`", () => {
  cli.start("post")
  expect(postGenerator).toBeCalledTimes(1)
})

test("Should run preview on `blog preview`", () => {
  cli.start("preview")
  expect(preview).toBeCalledTimes(1)
})

test("Should run fs publisher on `blog publish`", () => {
  cli.start("publish")
  expect(fsPublisher).toBeCalledTimes(1)
})
