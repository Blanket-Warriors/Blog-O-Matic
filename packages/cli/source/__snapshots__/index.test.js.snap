// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should log help text on no argument 1`] = `
"
Looks like you could use a little help!
Here are a few commands you can use...
  - \`blog init\`: Generate a blog in this directory
  - \`blog post\`: Generate a post in this blog
  - \`blog preview\`: Run a preview server of your blog
  - \`blog publish\`: Publish your changes!
"
`;
